cf-custom-resources/lib/custom-domain-app-runner.js

```diff
@@ -234,25 +241,128 @@ async function validateCertForDomain(serviceARN, domainName) {
+        // There are one known scenarios where status could be ACTIVE right after it's associated:
+        // When the domain just got deleted and added again. In this case, even though the validation records could
+        // have been deleted, the previously successful validation results are still cached. Because of the cache,
+        // the domain will show to be ACTIVE immediately after it's associated , although the validation records are not
+        // there anymore.
+        // In this case, the status won't transit to PENDING_VERIFICATION, so we need to check whether the validation
+        // records are ready by counting if there are three of them.
+        return true;
+    }
+
+    return false;
+}
+
+/**
+ * Remove custom domain from service by disassociating and removing the records for both the domain and the validation.
+ * If the custom domain is not found in the service, the function returns without error.
+ * Errors are not handled and are directly passed to the caller.
+ *
+ * @param {string} serviceARN ARN of the service that the custom domain applies to.
+ * @param {string} customDomainName the custom domain name.
+ */
+async function removeCustomDomain(serviceARN, customDomainName) {
+    let data;
+    try {
+        data = await appRunnerClient.disassociateCustomDomain({
+            DomainName: customDomainName,
+            ServiceArn: serviceARN,
+        }).promise();
+    } catch (err) {
+        if (err.message.includes(`No custom domain ${customDomainName} found for the provided service`)) {
+            return;
+        }
+        throw err;
+    }
+
+    return Promise.all([
+        updateCNAMERecordAndWait(customDomainName, data.DNSTarget, appHostedZoneID, "DELETE"), // Upsert the record that maps `customDomainName` to the DNS of the app runner service.
+        removeValidationRecords(data.CustomDomain),
+    ]);
+}
+
+/**
+ * Remove validation records for a custom domain.
+ *
+ * @param {object} domain information containing DomainName, Status, CertificateValidationRecords, etc.
+ * @throws wrapped error.
+ */
+async function removeValidationRecords(domain) {
+    const records = domain.CertificateValidationRecords;
+    let promises = [];
+    for (const record of records) {
+        promises.push(
+            updateCNAMERecordAndWait(record.Name, record.Value, appHostedZoneID, "DELETE").catch(err => {
+                throw new Error(`delete validation records for domain ${domain.DomainName}: ` + err.message);
+            })
+        );
+    }
+    return Promise.all(promises);
+}
+
+/**
+ * Wait for the custom domain to be disassociated.
+ * @param {string} serviceARN the service to which the domain is added.
+ * @param {string} customDomainName the domain name.
+ */
+async function waitForCustomDomainToBeDisassociated(serviceARN, customDomainName) {
+    let i, lastDomainStatus, domain;
+    for (i = 0; i < ATTEMPTS_WAIT_FOR_DISASSOCIATED; i++) {
```

Comment:
```suggestion
    let lastDomainStatus, domain;
    for (let i = 0; i < ATTEMPTS_WAIT_FOR_DISASSOCIATED; i++) {
```
Also looks like we are not using `domain` outside the loop